### ISSUE TITLE
Image cache doesn't persist tag information

### ISSUE THREAD

**Issue by fdawg4l** [vmware/vic#2064]
https://github.com/vmware/vic/blob/master/cmd/imagec/imagec.go#L391

```
385 »···log.Debugf("%d images require metadata updates", len(updatedImages))
386 »···for id := range updatedImages {
387 »···»···// utlizing index to avoid copy
388 »···»···update := updatedImages[id]
389 »···»···log.Debugf("updating ImageMetadata: layerID(%s)", update.ID)
390 »···»···//Send the Metadata to the PL to write to disk
391 »···»···err = WriteImage(update, nil)
```

`WriteImage` can only be called once for a given ID.  Calling it again with nil data will not update the metadata associated with an image.  So basically nothing is being persisted here.  

We need another means for you to keep an updated set of tag information which is mutable.  The Image calls won't allow you to do that (and for good reason-  images are immutable).

**Comment by fdawg4l** [comment]
```
374 »···»···»···»···»···meta: string(meta),
375 »···»···»···»···»···// create a dummy sum as this is required by the portlayer,
376 »···»···»···»···»···// but since we are only updating metadata it will be ignored
377 »···»···»···»···»···layer: FSLayer{BlobSum: "this-is-required"},
378 »···»···»···»···} 
```

Yeah, this is just wrong.

**Comment by fdawg4l** [comment]
cc: @hickeng

**Comment by jzt** [comment]
> We need another means for you to keep an updated set of tag information which is mutable. The Image calls won't allow you to do that (and for good reason- images are immutable).

Isn't this precisely what the PL-side k/v store was supposed to solve?

In docker world, the `repositories.json` file serves this purpose:

```json
{
  "Repositories": {
    "busybox": {
      "busybox:1.25.0": "sha256:2b8fd9751c4c0f5dd266fcae00707e67a2545ef34f9a29354585f93dac906749",
      "busybox:latest": "sha256:2b8fd9751c4c0f5dd266fcae00707e67a2545ef34f9a29354585f93dac906749",
      "busybox@sha256:a59906e33509d14c036c8678d687bd4eec81ed7c4b8ce907b888c607f6a1e0e6": "sha256:2b8fd9751c4c0f5dd266fcae00707e67a2545ef34f9a29354585f93dac906749"
    },
    "hello-world": {
      "hello-world:latest": "sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc",
      "hello-world@sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9": "sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc"
    }
  }
}
```

We discussed this with @corrieb and @hickeng when we were designing the approach for the image metadata/caching stuff.

**Comment by hickeng** [comment]
@jzt yes, this is specifically what the global PL k/v store was intended for.

**Comment by fdawg4l** [comment]
@hickeng Just closing the loop.  An issue asking for one was never created so it wasn't done.

 Created #2080 to track it.  Obviously, this is required for GA.

**Comment by mdubya66** [comment]
Moving to Icebox per 9/20 Triage

**Comment by fdawg4l** [comment]
We need a way to identify which layers in a `repository:tag` we can delete.

```
{
  "Repositories": {
    "busybox": {
      "busybox:1.25.0": "sha256:2b8fd9751c4c0f5dd266fcae00707e67a2545ef34f9a29354585f93dac906749",
      "busybox:latest": "sha256:2b8fd9751c4c0f5dd266fcae00707e67a2545ef34f9a29354585f93dac906749",
      "busybox@sha256:a59906e33509d14c036c8678d687bd4eec81ed7c4b8ce907b888c607f6a1e0e6": "sha256:2b8fd9751c4c0f5dd266fcae00707e67a2545ef34f9a29354585f93dac906749"
    },
    "hello-world": {
      "hello-world:latest": "sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc",
      "hello-world@sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9": "sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc"
    }
  }
}
```

```
vagrant@devbox /U/z/g/s/g/v/vic ❯❯❯ docker rmi butts
Untagged: butts:latest
vagrant@devbox /U/z/g/s/g/v/vic ❯❯❯ docker tag bf2b4c2d7bf5 nginx:butts
vagrant@devbox /U/z/g/s/g/v/vic ❯❯❯ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
nginx               1.10                bf2b4c2d7bf5        12 days ago         180.7 MB
nginx               butts               bf2b4c2d7bf5        12 days ago         180.7 MB
nginx               latest              ba6bed934df2        12 days ago         181.4 MB
vagrant@devbox /U/z/g/s/g/v/vic ❯❯❯ docker rmi bf2b4c2d7bf5
Error response from daemon: conflict: unable to delete bf2b4c2d7bf5 (must be forced) - image is referenced in one or more repositories
```

The `repository:tag` tuple represents a pointer in a tree of image layers.  If we delete "nginx:latest" via `rmi`, we need to delete all of the layers in the chain up to the next tag.

One option is to include the following in the image cache in the personality.

```
type Repository struct {
       // map tag name to ID
       Tags map[string]string
}

// map repo names to their Repository struct
Index = make(map[string]*Repository
```

When deleting any version of a given `Repository`, the storage layer should get a list of image ids for all of the other IDs for that given `Tag`.  If we can get a list of IDs of the other tags, we know where to stop deleting as we go up the chain in the storage layer.

**Comment by cgtexmex** [comment]
This issue has touched on a couple of different issues within the code base.  The original of not persisting the tag information was fixed.  It has subsequently been broken, but that fix and regression test will be tracked in #3025 

The other issue addressed here is the fact that vic only deletes the leaf layer of the images and that causes problems.  That issue is being tracked in #2973 

So closing this issue as it is now redundant on many levels.
